File: Python/2-TwoPointers/Problems/2-TwoSum-II(SortedArray)/TwoSum-II(SortedArray).py
# Two Sum II - Sorted Array Solutions
# Given a sorted array and a target sum, find two numbers that add up to the target.

# Approach 1: Two Pointers (Optimal Solution)
# Time Complexity: O(n), Space Complexity: O(1)
def two_sum_two_pointers(numbers, target):
    """
    Finds two numbers in a sorted array that add up to the target using the two-pointer approach.

    This is the most efficient solution because it takes advantage of the sorted nature of the array.
    It uses two pointers, one at the beginning and one at the end, and moves them towards each other
    based on whether the current sum is less than or greater than the target.

    Args:
        numbers: A sorted list of integers.
        target: The target sum.

    Returns:
        A list containing the 1-based indices of the two numbers that add up to the target.
        Returns an empty list if no such pair exists.
    """
    left, right = 0, len(numbers) - 1  # Initialize left and right pointers
    while left < right:  # Continue as long as the left pointer is before the right pointer
        current_sum = numbers[left] + numbers[right]  # Calculate the sum of the numbers at the two pointers
        if current_sum == target:
            return [left + 1, right + 1]  # 1-based index # If the sum equals the target, return the indices
        elif current_sum < target:
            left += 1  # If the sum is less than the target, move the left pointer to increase the sum
        else:
            right -= 1  # If the sum is greater than the target, move the right pointer to decrease the sum
    return []  # If no such pair is found, return an empty list

# Approach 2: Binary Search (Logarithmic Search for Complement)
# Time Complexity: O(n log n), Space Complexity: O(1)
def two_sum_binary_search(numbers, target):
    """
    Finds two numbers in a sorted array that add up to the target using binary search.

    This approach iterates through the array, and for each number, it uses binary search to
    find the complement (the number needed to reach the target) in the remaining part of the array.

    Args:
        numbers: A sorted list of integers.
        target: The target sum.

    Returns:
        A list containing the 1-based indices of the two numbers that add up to the target.
        Returns an empty list if no such pair exists.
    """
    for i in range(len(numbers)):  # Iterate through each number in the array
        complement = target - numbers[i]  # Calculate the complement needed
        left, right = i + 1, len(numbers) - 1  # Initialize left and right pointers for binary search
        while left <= right:  # Perform binary search
            mid = left + (right - left) // 2  # Calculate the middle index
            if numbers[mid] == complement:
                return [i + 1, mid + 1]  # If the complement is found, return the indices
            elif numbers[mid] < complement:
                left = mid + 1  # If the middle number is less than the complement, search in the right half
            else:
                right = mid - 1  # If the middle number is greater than the complement, search in the left half
    return []  # If no such pair is found, return an empty list

# Approach 3: Hash Map (Trade-off: More Space, Faster Lookup)
# Time Complexity: O(n), Space Complexity: O(n)
def two_sum_hash_map(numbers, target):
    """
    Finds two numbers in an array that add up to the target using a hash map.

    This approach uses a hash map (dictionary in Python) to store each number and its index.
    It then iterates through the array, checking if the complement of the current number
    is already in the hash map.  This provides a fast lookup (average O(1) time).

    Args:
        numbers: A list of integers.
        target: The target sum.

    Returns:
        A list containing the 1-based indices of the two numbers that add up to the target.
        Returns an empty list if no such pair exists.
    """
    seen = {}  # Initialize a hash map to store numbers and their indices
    for i, num in enumerate(numbers):  # Iterate through the array
        complement = target - num  # Calculate the complement needed
        if complement in seen:
            return [seen[complement] + 1, i + 1]  # If the complement is in the hash map, return the indices
        seen[num] = i  # Store the current number and its index in the hash map
    return []  # If no such pair is found, return an empty list

# Approach 4: Brute Force (Exhaustive Search)
# Time Complexity: O(n^2), Space Complexity: O(1)
def two_sum_brute_force(numbers, target):
    """
    Finds two numbers in an array that add up to the target using brute force.

    This is the simplest but least efficient approach.  It iterates through every possible
    pair of numbers in the array and checks if their sum equals the target.

    Args:
        numbers: A list of integers.
        target: The target sum.

    Returns:
        A list containing the 1-based indices of the two numbers that add up to the target.
        Returns an empty list if no such pair exists.
    """
    for i in range(len(numbers)):  # Iterate through each number in the array
        for j in range(i + 1, len(numbers)):  # Iterate through the remaining numbers
            if numbers[i] + numbers[j] == target:
                return [i + 1, j + 1]  # If the sum equals the target, return the indices
    return []  # If no such pair is found, return an empty list

# Approach 5: Recursive Two Pointers (Divide and Conquer)
# Time Complexity: O(n), Space Complexity: O(n) (Due to recursion stack)
def two_sum_recursive(numbers, target, left=0, right=None):
    """
    Finds two numbers in a sorted array that add up to the target using recursion
    and the two-pointer approach.

    This approach is similar to the iterative two-pointer approach, but it uses recursion
    instead of a loop.  It repeatedly divides the array and checks the sum of the
    numbers at the left and right pointers.

    Args:
        numbers: A sorted list of integers.
        target: The target sum.
        left: The left pointer (default: 0).
        right: The right pointer (default: len(numbers) - 1).

    Returns:
        A list containing the 1-based indices of the two numbers that add up to the target.
        Returns an empty list if no such pair exists.
    """
    if right is None:
        right = len(numbers) - 1  # Initialize the right pointer if it's not provided
    if left >= right:
        return []  # Base case: if the pointers cross, no solution is found
    current_sum = numbers[left] + numbers[right]  # Calculate the current sum
    if current_sum == target:
        return [left + 1, right + 1]  # If the sum equals the target, return the indices
    elif current_sum < target:
        return two_sum_recursive(numbers, target, left + 1, right)  # Recursively search in the right half
    else:
        return two_sum_recursive(numbers, target, left, right - 1)  # Recursively search in the left half

# Test Cases
numbers = [2, 7, 11, 15]
target = 9
print("Two Pointers:", two_sum_two_pointers(numbers, target))
print("Binary Search:", two_sum_binary_search(numbers, target))
print("Hash Map:", two_sum_hash_map(numbers, target))
print("Brute Force:", two_sum_brute_force(numbers, target))
print("Recursive Two Pointers:", two_sum_recursive(numbers, target))

File: C++/2-TwoPointers/Problems/2-TwoSum-II(SortedArray)/TwoSum-II(SortedArray).cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <tuple>
#include <stdexcept>

// Function to find two numbers in a sorted array that add up to a target sum.
//
// Args:
//     numbers: A sorted vector of integers.
//     target: The target sum.
//     approach: An integer specifying the approach to use (1 to 5).
//
// Returns:
//     A tuple containing the indices (1-based) of the two numbers if found,
//     otherwise a tuple with (-1, -1).
//
// Throws:
//     std::invalid_argument: If the approach is invalid.
std::tuple<int, int> twoSum(const std::vector<int>& numbers, int target, int approach) {
    if (numbers.empty()) {
        return std::make_tuple(-1, -1);
    }

    switch (approach) {
        case 1: {
            // Approach 1: Two Pointers (Optimal for Sorted Array)
            // Time Complexity: O(n)
            // Space Complexity: O(1)
            int left = 0;
            int right = numbers.size() - 1;
            while (left < right) {
                int sum = numbers[left] + numbers[right];
                if (sum == target) {
                    return std::make_tuple(left + 1, right + 1); // 1-based indexing
                } else if (sum < target) {
                    left++; // Move left pointer to increase sum
                } else {
                    right--; // Move right pointer to decrease sum
                }
            }
            return std::make_tuple(-1, -1); // Not found
        }
        case 2: {
            // Approach 2: Binary Search
            // Time Complexity: O(n log n)
            // Space Complexity: O(1)
            for (int i = 0; i < numbers.size(); ++i) {
                int complement = target - numbers[i];
                // Use binary search to find the complement in the remaining part of the array
                int left = i + 1;
                int right = numbers.size() - 1;
                while (left <= right) {
                    int mid = left + (right - left) / 2; // Prevent overflow
                    if (numbers[mid] == complement) {
                        return std::make_tuple(i + 1, mid + 1); // 1-based indexing
                    } else if (numbers[mid] < complement) {
                        left = mid + 1;
                    } else {
                        right = mid - 1;
                    }
                }
            }
            return std::make_tuple(-1, -1); // Not found
        }
        case 3: {
            // Approach 3: Hash Map (unordered_map)
            // Time Complexity: O(n) on average, O(n^2) worst-case
            // Space Complexity: O(n)
            std::unordered_map<int, int> numMap; // Key: number, Value: index
            for (int i = 0; i < numbers.size(); ++i) {
                int complement = target - numbers[i];
                if (numMap.find(complement) != numMap.end()) {
                    return std::make_tuple(numMap[complement] + 1, i + 1); // 1-based indexing
                }
                numMap[numbers[i]] = i;
            }
            return std::make_tuple(-1, -1); // Not found
        }
        case 4: {
            // Approach 4: Brute Force
            // Time Complexity: O(n^2)
            // Space Complexity: O(1)
            for (int i = 0; i < numbers.size(); ++i) {
                for (int j = i + 1; j < numbers.size(); ++j) {
                    if (numbers[i] + numbers[j] == target) {
                        return std::make_tuple(i + 1, j + 1); // 1-based indexing
                    }
                }
            }
            return std::make_tuple(-1, -1); // Not found
        }
        case 5: {
            // Approach 5: Using std::find (Illustrative, not optimal)
            // Time Complexity: O(n^2) in worst case.  std::find is O(n)
            // Space Complexity: O(1)
             for (int i = 0; i < numbers.size(); ++i) {
                int complement = target - numbers[i];
                //  std::find  iterates through the *entire* remaining portion.
                auto it = std::find(numbers.begin() + i + 1, numbers.end(), complement);
                if (it != numbers.end()) {
                    int j = std::distance(numbers.begin(), it); // Get the index
                    return std::make_tuple(i + 1, j + 1);
                }
             }
            return std::make_tuple(-1, -1);
        }
        default:
            throw std::invalid_argument("Invalid approach specified.  Choose 1-5.");
    }
}

int main() {
    // Example usage
    std::vector<int> numbers = {2, 7, 11, 15};
    int target = 9;

    // Test each approach
    for (int approach = 1; approach <= 5; ++approach) {
        std::tuple<int, int> result = twoSum(numbers, target, approach);
        int index1, index2;
        std::tie(index1, index2) = result; // Unpack the tuple

        std::cout << "Approach " << approach << ": ";
        if (index1 != -1 && index2 != -1) {
            std::cout << "Indices: " << index1 << ", " << index2 << std::endl;
        } else {
            std::cout << "No solution found." << std::endl;
        }
    }

    // Example with no solution
    std::vector<int> numbers2 = {2, 3, 4, 5};
    target = 12;
    std::tuple<int, int> result = twoSum(numbers2, target, 1); // Using approach 1
    int index1, index2;
    std::tie(index1, index2) = result;
    std::cout << "No Solution Example: Approach 1: ";
     if (index1 != -1 && index2 != -1) {
        std::cout << "Indices: " << index1 << ", " << index2 << std::endl;
    } else {
        std::cout << "No solution found." << std::endl;
    }
    // Example with an empty vector
    std::vector<int> numbers3 = {};
    target = 0;
    std::tuple<int, int> resultEmpty = twoSum(numbers3, target, 1);
    int index1Empty, index2Empty;
    std::tie(index1Empty, index2Empty) = resultEmpty;
    std::cout << "Empty Vector Example: Approach 1: ";
    if (index1Empty != -1 && index2Empty != -1) {
        std::cout << "Indices: " << index1Empty << ", " << index2Empty << std::endl;
    } else {
        std::cout << "No solution found." << std::endl;
    }

    // Example with duplicate numbers
    std::vector<int> numbers4 = {3, 3, 6, 15};
    target = 6;
    std::tuple<int, int> resultDuplicate = twoSum(numbers4, target, 1);
    int index1Duplicate, index2Duplicate;
    std::tie(index1Duplicate, index2Duplicate) = resultDuplicate;
    std::cout << "Duplicate Numbers Example: Approach 1: ";
    if (index1Duplicate != -1 && index2Duplicate != -1) {
        std::cout << "Indices: " << index1Duplicate << ", " << index2Duplicate << std::endl;
    } else {
        std::cout << "No solution found." << std::endl;
    }
    return 0;
}
